brave/brave-core
# Improve Brave ads serving pipeline performance

Resolves https://github.com/brave/brave-browser/issues/18843

## Submitter Checklist:


## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on

## Test Plan:

As a minimum QA should test areas highlighted in **BOLD**. Other areas were unit tested before the changes. For performance tests: Join rewards (for United States), Quit browser, Replace `database.sqlite`, Launch browser and then serve an ad (monitor performance in the console log).

Regression test Brave ads including:
- **user should be rewarded for viewing the following types of ad:**
  - push notification ads
  - inline content ads
  - promoted content ads
  - new tab page ads
- confirm ads are served for:
  - push notification ads for v1 pipeline (parent-child, parent and untargeted segments)
  - push notification ads for v2 pipeline
  - **Confirm CPU usage for 4,200+ entries in the `ad_events` database table and a large catalog, i.e. United States (QA please reach out via DM for test file. NOTE: entries are purged after 3 months) for push notification ads v1 pipeline**
  - **Confirm CPU usage for 4,200+ entries in the `ad_events` database table and a large catalog, i.e. United States (QA please reach out via DM for test file. NOTE: entries are purged after 3 months) for push notification ads v2 pipeline**
  - inline content ads for v1 pipeline (untargeted segments)
  - inline content ads for v2 pipeline
- frequency caps working as expected for:
  - anti-targeting (capped against creative set)
  - conversion (capped against creative set)
  - daily cap (capped against campaign)
  - day-parting (capped against creative set)
  - dislike (thumbs down icon in ads history, capped against creative set)
  - dismissed (user closed push notification ad, capped against campaign)
  - marked as inappropriate (via kebab menu in ads history, capped against creative set)
  - marked to no longer receive (no entry icon in ads history, capped against creative set)
  - per day (capped against creative set)
  - per hour (capped against creative instance)
  - per month (capped against creative set)
  - per week (capped against creative set)
  - split test (capped against creative set)
  - subdivision targeting (capped against creative set)
  - total max (capped against creative set)
  - transferred (aka Landed, capped against campaign)
- Ad pacing (`ptr` in catalog)
- Ad prioritization (`priority` in catalog)
- Upgrade path from 1.33.30 (database schema 16) to 1.33.31 or above (database schema 17)

NOTE: Frequency caps were more detailed in the diagnostic logs prior to this change, however going forward we will not see a complete list of reasons why ads were capped due to the caching of already capped campaign, advertiser, creative set and creative instance ids to reduce the amount of business logic we run. i.e., if an ad was capped due to creative set any other ad capped for the same creative set would also be capped but not logged.
